### PR TITLE
Activity worker and godeps and  testing helpers

### PR DESCRIPTION
**Pull request by sclasen**
/cc @fabiokung @ryanbrainard @dgouldin @dpiddy

this is some initial work on an ActivityWorker which brings similar things to the Activity side of things that FSM brings to the decision side of things.

One main motivation here is to be easily able to provide mock activity workers that  just return data based on their inputs without doing any work, so we can exercise FSMs more in actual against swf testing. These mocks can be written to fail most of the time so we can exercise failure scenarios well in FSMs.

Will point out some relevant areas in comments below, this is still very much WIP.

TODO:  abstract out long-lived/heartbeating activity workers and fit them in here, or in a seperate LongLivedActivityWorker abstraction.

### CODE REVIEW

File: activity/worker_test.go [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+	Data2 string
+}
+
+type Output2 struct {
+	Data2 string
+}
+
+//We define the operations our activity worker will handle in an interface
+//Then it is easy to provide a mocked impl
+type Activities interface {
```
Comment (sclasen) [inline activity/worker_test.go]: This is an example Interface of an activity worker that handles one or more activity types.

File: activity/worker_test.go [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+	Task1(*swf.ActivityTask, *Input1) (*Output1, error)
+	Task2(*swf.ActivityTask, *Input2) (*Output2, error)
+}
+
+type MockActivities struct{}
+
+func (m MockActivities) Task1(task *swf.ActivityTask, in *Input1) (*Output1, error) {
```
Comment (sclasen) [inline activity/worker_test.go]: Here we provide a mock impl that fails 60% of the time

File: activity/worker_test.go [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+
+	worker := &ActivityWorker{
+		Domain:     domain,
+		Serializer: fsm.JSONStateSerializer{},
+		TaskList:   taskList,
+		SWF:        client,
+		Identity:   "test-activity-worker",
+	}
+
+    //This is where the actual worker config code is
```
Comment (sclasen) [inline activity/worker_test.go]: Here is how we wire up the worker.  the machinery takes care of all the serialization/deserialization, calling done/fail, etc.

File: activity/worker.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+func (a *ActivityWorker) AddHandler(handler *ActivityHandler) {
+	if a.handlers == nil {
+		a.handlers = map[string]*ActivityHandler{}
+	}
+	a.handlers[handler.Activity] = handler
+}
+
+func (a *ActivityWorker) Init() {
```
Comment (fabiokung) [inline activity/worker.go]: I'd much rather see an `activity.New(required...) *activityActivityWorker` which sets defaults for all fields (e.g.: `exec.Command(...) *exec.Cmd`).

**Comment by fabiokung** [comment]
Good work! The overall strategy lgtm.

I'm not particularly happy with all the reflection in `activity.NewActivityHandler`, but I can't suggest anything better though. It's just Go's type system not being good enough.

File: activity/dispatchers.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
+package activity
+
+import "github.com/awslabs/aws-sdk-go/gen/swf"
+
+//ActivityTaskDispatcher is used by the FSM machinery to
```
Comment (ryanbrainard) [inline activity/dispatchers.go]: `used by the FSM machinery to...`?

Comment (sclasen) [inline activity/dispatchers.go]: hehe cut and paste there, will update.

**Comment by sclasen** [comment]
Gonna bring this in now, open new issues for long-lived activity worker abstraction, and documentation on testing code

**Pull request merged**